feat(test): OOG and success selfdestruct tests

fselmo opened the pull request:
## 🗒️ Description

Added comprehensive `SELFDESTRUCT` tests covering beneficiaries as EOA, contract, precompiles, and self. All tests consolidated in `TangerineWhistle` with fork-aware parametrization

  - is_success (exact_gas / exact_gas_minus_1) - tests success boundary
  - warm param uses `pytest.param(..., marks=valid_from("Berlin"))` so warm tests only run on `>=Berlin`
  - State access boundary tests (consensus check: beneficiary accessed before `G_NEW_ACCOUNT` evaluated)
  - same_tx vs pre-deployed contracts
  - originator_balance (0/1) - EIP-161 behavior difference
  - beneficiary_initial_balance (0/1) - dead/alive for `G_NEW_ACCOUNT`

Precompile coverage was sparse - some static tests hit 0x01, others 0x01-0x03. Now using `@pytest.mark.with_all_precompiles` for full coverage across all fork precompiles.

TangerineWhistle (EIP-150): All tests live here since this is when gas costs for `SELFDESTRUCT` were introduced. Fork-aware parametrization handles Berlin+ warm/cold distinction.

SpuriousDragon (EIP-161, EIP-155): Tests correctly handle the originator_balance > 0 requirement for G_NEW_ACCOUNT. ~~Also updated old tests to use >= SpuriousDragon for protected transactions since this is when EIP-155 was introduced, not Byzantium as other tests have been using in our codebase.~~ [This was then refactored as `fork.supports_protected_txs`, returning `True` beginning at `SpuriousDragon`].

Berlin (EIP-2929): Warm/cold coverage integrated via warm param with `valid_from("Berlin")` mark - no separate Berlin test file needed.

## 🔗 Related Issues or PRs

- Related to a Nethermind bug in BAL devnet 1, relevant PR [here](https://github.com/NethermindEth/nethermind/pull/10064)

## ✅ Checklist
<!-- Please check off all required items. For those that don't apply remove them accordingly. -->

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [x] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.

#### Cute Animal Picture

<img width="1046" height="770" alt="Screenshot 2025-12-29 at 14 19 44" src="https://github.com/user-attachments/assets/e2bb87db-1447-4394-bda9-14d8e949411d" />

fselmo committed:
feat(test): OOG and success selfdestruct tests to all precompiles

feat(test): Expand fork range for selfdestruct to precompile tests

- For successful tests, start at Homestead where precompiles were
  introduced (EIPs 196, 197, 198).

- For OOG tests, start at Tangerine where operation gas costs were
  introduced (EIP 150).

fselmo commented:
I think we should move these to an earlier fork as it looks like we may not have coverage for this. I see _some_ coverage in the static tests but not for all opcodes. It would be good to have this in python either way and use the `pytest.mark.with_all_precompiles` to test every added precompile for every fork as they are added.

I'll work on porting this to an earlier fork.

jochem-brouwer commented:
Does this initialize the precompiles as empty and does it attempt to create the accounts via `SELFDESTRUCT`? 

This is definitely a case which should be added, but just wanted to note here that due to this bug:

<img width="1121" height="155" alt="image" src="https://github.com/user-attachments/assets/7f3de57b-1844-45b8-a1fb-5b598fa16d8f" />

(from the yellow paper) there is this weird behavior with precompiles, creating the account, the precompile going out-of-gas and then the account cleanup either cleaning up the account (or not). 

This bug caused that later devnets and all live chains just funded the addresses with 1 wei, such that the account is always there (and not the "0 balance, 0 nonce, no code, no storage", but still in trie case, which now is not possible anymore since it has a balance)

I wanted to share this because this might be the reason why the coverage report is low. I think (although this should be an EIP in that case) that the general consensus is to "just fund the precompiles" with 1 wei at genesis to get rid of this behavior, and that it is thus either unspecified or untested behavior.

fselmo commented:
> Does this initialize the precompiles as empty and does it attempt to create the accounts via SELFDESTRUCT?

`pytest.mark.with_all_precompiles` will parametrize for all precompile addresses that are mapped for the fork you are filling for, but it doesn't add anything to the `pre`. All the EVM knows is that this is the beneficiary for the self-destruct. Is there a missing case? 🤔

jochem-brouwer commented:
> `pytest.mark.with_all_precompiles` will parametrize for all precompile addresses that are mapped for the fork you are filling for, but it doesn't add anything to the `pre`. All the EVM knows is that this is the beneficiary for the self-destruct. Is there a missing case? 🤔

Ah sorry, I think my comment is confusing. The YP case is, IIRC, a case which is only possible if a precompile account is empty (no balance, 0 nonce, no code, no storage). I am not sure but I thought some tests pre-fill precompiles with a balance to make them non-empty, and wanted to explicitly mention this here. But I think this is already the case. So no missing cases found here.

jochem-brouwer commented on packages/testing/src/execution_testing/forks/forks/forks.py:
```diff
@@ -947,7 +947,7 @@ class SpuriousDragon(Tangerine, ignore=True):
     pass
```
Could you edit the description of the Spurious Dragon fork? It lists EIP-158 to be in there but this is not the case (this was a non-shipped variant of EIP-161).

Ref: https://eips.ethereum.org/EIPS/eip-607

jochem-brouwer commented on packages/testing/src/execution_testing/forks/forks/forks.py:
```diff
@@ -935,7 +935,7 @@ class DAOFork(Homestead, ignore=True):
 
 
-class Tangerine(DAOFork, ignore=True):
+class Tangerine(DAOFork, transition_tool_name="TangerineWhistle"):
```
Why is SpuriousDragon called SpuriousDragon and TangerineWhistle Tangerine?

jochem-brouwer commented on packages/testing/src/execution_testing/forks/__init__.py:
```diff
@@ -97,6 +98,7 @@
     "Frontier",
     "GrayGlacier",
     "Homestead",
+    "Tangerine",
```
If this is the "official" fork name then this should be TangerineWhistle.

fselmo replied:
Yes I agree. I just turned this on (it was already defined). I'm happy to change it here though as I don't think that would complicate anything since it is currently ignored.

jochem-brouwer commented on tests/amsterdam/eip7928_block_level_access_lists/test_cases.md:
```diff
@@ -105,11 +105,10 @@
 | `test_bal_lexicographic_address_ordering` | Ensure BAL enforces strict lexicographic byte-wise ordering | Pre-fund three addresses with specific byte patterns: `addr_low = 0x0000...0001`, `addr_mid = 0x0000...0100`, `addr_high = 0x0100...0000`. Contract touches them in reverse order: `BALANCE(addr_high), BALANCE(addr_low), BALANCE(addr_mid)`. Additionally, include two endian-trap addresses that ...
 | `test_bal_transient_storage_not_tracked` | Ensure BAL excludes EIP-1153 transient storage operations | Contract executes: `TSTORE(0x01, 0x42)` (transient write), `TLOAD(0x01)` (transient read), `SSTORE(0x02, result)` (persistent write using transient value). | BAL **MUST** include slot 0x02 in `storage_changes` (persistent storage was modified). BAL **MUST NOT** include slot 0x01 in `storage_re ...
-| `test_bal_selfdestruct_to_precompile` | Ensure BAL captures SELFDESTRUCT with precompile as beneficiary | Caller triggers victim contract (balance=100) to execute `SELFDESTRUCT(0x0000...0001)` (ecrecover precompile). Precompile starts with balance=0. | BAL **MUST** include: (1) Contract with `balance_changes` (100→0, loses balance to selfdestruct). (2) Precompile address 0x01 with `balance_chan ...
+| `test_bal_selfdestruct_to_precompile_and_oog` | Ensure BAL captures SELFDESTRUCT to precompile at different gas boundaries | Victim executes `SELFDESTRUCT(precompile)`. Parameterized by all precompiles and three scenarios: (1) Success, (2) OOG before state access, (3) OOG after state access. | Success: victim and precompile have `balance_changes`. OOG before state access: precompile **NOT** in  ...
```
(3) is only possible in this case if we send value, right?

fselmo replied:
If we have >= `selfdestruct (5000)` cost, we reach into the `accessed_addresses` for `beneficiary` to see if it's alive or not and so we mark it warm. It's a bit of a pedantic case where it may be marked "warm" even though we don't _need_ to touch it (if `0` transfer). But as far as the access list is concerned, whether we touch it in state or not with optimizations, this means it's already warmed at this point and is in BAL. The EIP mentions this specifically in the [Scope and Inclusion](https://eips.ethereum.org/EIPS/eip-7928#scope-and-inclusion) section, under "addresses accessed without state changes, including":

> Beneficiary addresses for SELFDESTRUCT

jochem-brouwer replied:
So the gas checks are: `base cost (5000)` + `warm/cold access` + `create account`.

Just iterating this for myself but there are thus 4 boundaries:

1. Not enough gas to pay base cost
2. Not enough gas to pay in case target is cold (to warm it up) (If target is warm, this check costs 0)
3. Not enough gas to pay for creating the account
4. Success

So one of the cases is missing here :thinking:

jochem-brouwer replied:
Also linking EIP-2929: the warm/cold check is there even if the value of SELFDESTRUCT is 0 https://eips.ethereum.org/EIPS/eip-2929#selfdestruct-changes

fselmo replied:
Yes! This is missing some cases for sure thank you. I was thinking about this from a BAL perspective as this was a BAL test ported to earlier forks (pre state access and post state access). But since we are missing coverage for this, we should be more thorough on these boundaries in general with these tests. Thanks!

fselmo replied:
I have greatly extended the test cases now. Thanks!

jochem-brouwer reviewed (commented):
Not a review of the tests, just a small point regarding fork names and docs 😄 👍

fselmo committed:
refactor(test): `Tangerine` -> `TangerineWhistle`; comments on PR #1954

Bonus:

- fix(test): remove unnecessary isolation for enginex

fselmo committed:
fix(tool): Fix EvmOneTransitionTool parsing for TangerineWhistle (add space)

fselmo committed:
feat(test): Extend selfdestruct tests to all Amsterdam gas boundaries

fselmo commented:
@danceratopz this is ready for review now! Note that `json_infra` and `optimized` CI runs won't pass until the next bal test release is out and we point to it in the CI (blob params from the last release are at Osaka but the repo is now at BPO2).

fselmo commented on packages/testing/src/execution_testing/client_clis/transition_tool.py:
```diff
@@ -326,13 +327,19 @@ def _evaluate_filesystem(
         output_paths["body"] = os.path.join("output", "txs.rlp")
 
+        # Get fork name and apply any tool-specific mapping
+        fork_name = (
```
This fixed CI against EvmOne btw... for both forks that have a space in between. Even though I re-ignored them, I think this is a nice addition since it would just take turning the ignore on / off to turn these forks on again.

fselmo commented:
One thing here we'd have to eventually check is I believe this may cover a lot of cases that static tests currently cover. We can maybe, as a separate task, go through and check which tests are now covered here that we can possibly remove from `tests/static/`.

fselmo committed:
refactor(test): fork.supports_protected_txs() instead of direct comparison

- This will also help with support for any forks of this repo for L2s
  or other projects if some fork definitions don't have any meaningful
  relationship to SpuriousDragon.

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+            address=factory_address,
+            code=factory_code,
+            balance=originator_balance,
+        )
```
```suggestion
        factory_code = Op.EXTCODECOPY(
            initcode_address, 0, 0, initcode_len
        ) + Op.CALL(
            gas=inner_call_gas,
            address=Op.CREATE(
                value=originator_balance, offset=0, size=initcode_len
            ),
        )
        caller = pre.deploy_contract(
            code=factory_code,
            balance=originator_balance,
        )
        victim = compute_create_address(address=caller, nonce=1)
```
Should avoid the need to use `pre._contract_address_iterator` since this is not available in `execute`.

fselmo replied:
Ah, ty! I was hoping to not have to reach into the private API 👍🏼

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+        initcode_len = len(initcode)
+
+        # Pre-deploy initcode at separate address, then use EXTCODECOPY
+        initcode_address = pre.deploy_contract(initcode)
```
`MSTORE` macro should allow us to skip this contract deployment:

```
from execution_testing import Macros as Om
```
```
factory_code = Om.MSTORE(initcode, 0) + ...
```

fselmo replied:
I learn some new utility every day it seems! Thanks, this is really cool 👍🏼

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+    ids=["dead_beneficiary", "alive_beneficiary"],
+)
+@pytest.mark.valid_from("TangerineWhistle")
+def test_selfdestruct_to_precompile(
```
We could add system contracts as another test, I think they are also always warm iirc.

fselmo replied:
added here [518c5c8](https://github.com/ethereum/execution-specs/pull/1954/commits/518c5c8797f31d6f4fc19aa50dbde89c0f06ab76)

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+
+    Key characteristics:
+    - Beneficiary is always warm (it's the executing contract)
+    - Beneficiary is always alive (has code)
```
I think this does not hold in the case where self is a contract mid-creation (Initcode self-destructs).
We need to add such case.

fselmo replied:
I think it's worth it to add this but EIP-161 sets the nonce to `1` before the initcode runs ([section a](https://eips.ethereum.org/EIPS/eip-161#specification)). So I think in any case it's not really possible to test the "dead" case. I suppose the "has code" part doesn't hold, is this what you mean? I can make that more clear here and add this  test case as well 👍🏼.

marioevz replied:
Yeah, just the "has code" part is the thing that came to mind for me, it would not hurt to have it implemented, thanks!

fselmo replied:
added here ([518c5c8](https://github.com/ethereum/execution-specs/pull/1954/commits/518c5c8797f31d6f4fc19aa50dbde89c0f06ab76))

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+
+        initcode_address = pre.deploy_contract(initcode)
+
+        factory_address = next(pre._contract_address_iterator)  # type: ignore
```
Same recommendation as prior comment.

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+                                new_code=bytes(victim_code),
+                            )
+                        ],
+                    )
```
I think doing `victim_expectation.balance_changes.append(BalBalanceChange(...` might make it a bit easier to read instead of repeating almost the same code in two different blocks.

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1004 @@
+                    nonce_changes=[
+                        BalNonceChange(block_access_index=1, post_nonce=2)
+                    ],
+                    balance_changes=[
```
Same comment as above.

marioevz reviewed (commented):
Looks great! I made a couple suggestions, thanks!

fselmo committed:
refactor: address comments from PR #1954

fselmo committed:
feat(test): Add selfdestruct to system contracts + to self from initcode

fselmo committed:
feat(test-commands): Allow pytest valid fork markers as params; add unit tests

fselmo commented:
@marioevz Added changes from the comments here. I also realized that the `valid_from("Berlin")` pytest marker on the param was not working as expected. I added a `pytest_collection_modifyitems` to the forks plugin in order to sort the markers better. It worked out well but I also added some unit tests to make sure it's sanity checked. Lmk your thoughts on that approach. It makes it so we can run `warm` tests from Berlin on but `cold`, for example, for `TangerineWhistle` on. It's a nice added flexibility imo, which allowed me to not copy the entirety of the test over to Berlin just to add warm / cold paths.

gurukamath commented on src/ethereum/forks/amsterdam/vm/instructions/system.py:
```diff
@@ -640,7 +639,8 @@ def selfdestruct(evm: Evm) -> None:
         evm.accessed_addresses.add(beneficiary)
-        track_address(evm.state_changes, beneficiary)
+
+    track_address(evm.state_changes, beneficiary)
```
This is related to line 669 but leaving a comment here.

Could we defer the balance tracking for the `originator` and `beneficiary` until after the `same_tx` check on line 685? That way we do not have to balance track the `originator` twice for `same_tx`

fselmo replied:
This is here because we reach into the account in the very next line `is_account_alive()`. So this is not for the balance change, this is for checking if the account is alive. Does that make sense? I think functionally we could defer it, but semantically it makes sense to track it here.

gurukamath commented on src/ethereum/forks/amsterdam/vm/instructions/environment.py:
```diff
@@ -494,12 +494,12 @@ def extcodehash(evm: Evm) -> None:
     check_gas(evm, access_gas_cost)
     if is_cold_access:
         evm.accessed_addresses.add(address)
-    track_address(evm.state_changes, address)
```
For `balance`, `extcodesize`, `extcodecopy` and `extcodehash`, we don't need to call `check_gas` first and then call `charge_gas`. Could we directly call `charge_gas` in these cases?

fselmo replied:
Yes, I agree on this. This was part of the initial implementation of `check_gas`. I haven't looked back at those since then but I think this makes more sense for sure. Are you good with changing that here?

fselmo replied:
Added in latest commit [here](https://github.com/ethereum/execution-specs/pull/1954/commits/6cb77129260ced9d711899953d8ba8cc6591c0bd).

fselmo committed:
refactor(spec): Remove check_gas where unnecessary for instructions

marioevz commented on tests/tangerine_whistle/eip150_operation_gas_costs/test_eip150_selfdestruct.py:
```diff
@@ -0,0 +1,1197 @@
+    - No G_NEW_ACCOUNT charge (contract has nonce > 0)
+    - No cold access charge (>=Berlin)
+
+    Note: Gas boundary testing not possible for initcode since CREATE
+    doesn't accept a gas parameter - it uses all available gas.
```
See example in this PR: https://github.com/ethereum/execution-specs/pull/2002/files#diff-095178dc2c6d121067cadab23cb09943682751281a0489a73f90906e26855a6a

Basically we limit the gas passed to the factory contract, taking into account the 63/64 rule.

fselmo replied:
Really nice use 😄. I created #2010 to track this 👍🏼

marioevz reviewed (approved):
Approving based on:
- Tracking issue #2010 will fix outstanding comment.
- CI failures require new BAL release.
